packages/vm/lib/state/stateManager.ts

```diff
@@ -642,9 +661,57 @@ export default class DefaultStateManager implements StateManager {
 
   /**
    * Clear the warm accounts and storage. To be called after a transaction finished.
+   * @param boolean - If true, returns an EIP-2930 access list generated
    */
   clearWarmedAccounts(): void {
     this._accessedStorage = [new Map()]
+    this._accessedStorageReverted = [new Map()]
+  }
+
+  /**
+   * Generates an EIP-2930 access list
+   *
+   * Note: this method is not yet part of the `StateManager` interface.
+   * If not implemented, `runTx()` is not allowed to be used with the
+   * `reportAccessList` option and will instead throw.
+   *
+   * Note: there is an edge case on accessList generation where an
+   * internal call might revert without an accessList but pass if the
+   * accessList is used for a tx run (so the subsequent behavior might change).
+   * This edge case is not covered by this implementation.
+   *
+   * @returns - an [@ethereumjs/tx](https://github.com/ethereumjs/ethereumjs-monorepo/packages/tx) `AccessList`
+   */
+  generateAccessList(): AccessList {
+    // Merge with the reverted storage list
+    const mergedStorage = [...this._accessedStorage, ...this._accessedStorageReverted]
+
+    // Fold merged storage array into one Map
+    while (mergedStorage.length >= 2) {
+      const storageMap = mergedStorage.pop()
+      if (storageMap) {
+        this._accessedStorageMerge(mergedStorage, storageMap)
+      }
+    }
+    const folded = new Map([...mergedStorage[0].entries()].sort())
+
+    // Transfer folded map to final structure
+    const accessList: AccessList = []
+    folded.forEach((slots, addressStr) => {
+      const address = Address.fromString(`0x${addressStr}`)
+      if (!address.isPrecompileOrSystemAddress()) {
+        const storageSlots = Array.from(slots)
+          .map((s) => `0x${s}`)
+          .sort()
```

Comment:
Sorting this is fine, but it is not really necessary, right?

Reply:
I would be a fan of sorting (even if it comes to some performance costs), I think this gives a better overview if one want to manually check the outcome. And I guess in this use case performance is a bit secondary (just an unproven assumption reflecting on how I imagine people using this stuff).